pretyping/cases.ml

```diff
@@ -1055,11 +1055,7 @@ let add_assert_false_case pb tomatch =
 let adjust_impossible_cases sigma pb pred tomatch submat =
   match submat with
   | [] ->
-    (* FIXME: This breaks if using evar-insensitive primitives. In particular,
-       this means that the Evd.define below may redefine an already defined
-       evar. See e.g. first definition of test for bug #3388. *)
-    let pred = EConstr.Unsafe.to_constr pred in
-    begin match Constr.kind pred with
+    begin match EConstr.kind sigma pred with
     | Evar (evk, _) ->
       let EvarInfo evi = Evd.find sigma evk in
       if snd (Evd.evar_source evi) == Evar_kinds.ImpossibleCase then
```

Comment:
You can simplify the if-then-else clause below because now evk is always undefined.

Reply:
changed